File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/messages/ConnectionInvitation.java
package msdk.java.messages;

import android.webkit.URLUtil;

import com.evernym.sdk.vcx.VcxException;
import com.evernym.sdk.vcx.utils.UtilsApi;

import org.json.JSONObject;

import java.util.concurrent.ExecutionException;

import msdk.java.handlers.Connections.ConnectionMetadata;
import msdk.java.types.AriesMessageType;
import msdk.java.utils.CommonUtils;

public class ConnectionInvitation {
    public enum InvitationType {
        Connection,
        OutOfBand
    }

    public static boolean isAriesConnectionInvitation(ConnectionInvitation.InvitationType type) {
        return type == ConnectionInvitation.InvitationType.Connection;
    }

    public static boolean isAriesOutOfBandConnectionInvitation(ConnectionInvitation.InvitationType type) {
        return type == ConnectionInvitation.InvitationType.OutOfBand;
    }

    public static String getConnectionInvitationFromData(String data) {
        if (URLUtil.isValidUrl(data)) {
            try {
                return UtilsApi.vcxResolveMessageByUrl(data).get();
            } catch (InterruptedException | ExecutionException | VcxException exception) {
                exception.printStackTrace();
                return null;
            }
        } else {
            return data;
        }
    }

    public static ConnectionMetadata extractUserMetaFromInvitation(String invite) {
        JSONObject json = CommonUtils.convertToJSONObject(invite);
        if (json == null) {
            return new ConnectionMetadata("Unknown", "");
        }
        String label = json.optString("label");
        String logo = json.optString("profileUrl");
        return new ConnectionMetadata(label, logo);
    }

    public static InvitationType getInvitationType(String invite) {
        try {
            JSONObject json = new JSONObject(invite);
            String invitationType = json.optString("@type");
            if (invitationType.contains(AriesMessageType.OUTOFBAND_INVITATION)) {
                return InvitationType.OutOfBand;
            }
            if (invitationType.contains(AriesMessageType.CONNECTION_INVITATION)) {
                return InvitationType.Connection;
            }
            throw new Exception("Invalid invite format");
        } catch (Exception e ) {
            e.printStackTrace();
        }
        return null;
    }

    public static boolean compareInvites(String newInvite, String storedInvite) throws Exception {
        JSONObject newJson = new JSONObject(newInvite);
        JSONObject storedJson = new JSONObject(storedInvite);
        String newPublicDid = newJson.optString("public_did");
        String storedPublicDid = storedJson.optString("public_did");
        if (!storedPublicDid.isEmpty()) {
            return storedPublicDid.equals(newPublicDid);
        } else {
            String newDid = newJson.getJSONArray("recipientKeys").getString(0);
            String storedDid = storedJson.getJSONArray("recipientKeys").optString(0);
            return storedDid.equals(newDid);
        }
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/messages/ConnectionInvitation.kt
package msdk.kotlin.sample.messages

import android.webkit.URLUtil
import com.evernym.sdk.vcx.utils.UtilsApi
import msdk.kotlin.sample.handlers.Connections.ConnectionMetadata
import msdk.kotlin.sample.types.AriesMessageType
import msdk.kotlin.sample.utils.CommonUtils
import org.json.JSONObject
import java.util.concurrent.ExecutionException

object ConnectionInvitation {
    fun isAriesConnectionInvitation(type: InvitationType): Boolean {
        return type == InvitationType.Connection
    }

    fun isAriesOutOfBandConnectionInvitation(type: InvitationType): Boolean {
        return type == InvitationType.OutOfBand
    }

    fun getConnectionInvitationFromData(data: String): String {
        return if (URLUtil.isValidUrl(data)) {
            try {
                UtilsApi.vcxResolveMessageByUrl(data).get();
            } catch (ex: ExecutionException) {
                ex.printStackTrace()
                data
            } catch (ex: InterruptedException) {
                ex.printStackTrace()
                data
            }
        } else {
            data
        }
    }

    fun extractUserMetaFromInvitation(invite: String?): ConnectionMetadata {
        val json =
            CommonUtils.convertToJSONObject(invite) ?: return ConnectionMetadata("Unknown", "")
        val label = json.optString("label")
        val logo = json.optString("profileUrl")
        return ConnectionMetadata(label, logo)
    }

    fun getInvitationType(invite: String?): InvitationType? {
        try {
            val json = JSONObject(invite)
            val invitationType = json.optString("@type")
            if (invitationType.contains(AriesMessageType.OUTOFBAND_INVITATION)) {
                return InvitationType.OutOfBand
            }
            if (invitationType.contains(AriesMessageType.CONNECTION_INVITATION)) {
                return InvitationType.Connection
            }
            throw Exception("Invalid invite format")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    @Throws(Exception::class)
    fun compareInvites(newInvite: String?, storedInvite: String?): Boolean {
        val newJson = JSONObject(newInvite)
        val storedJson = JSONObject(storedInvite)
        val newPublicDid = newJson.optString("public_did")
        val storedPublicDid = storedJson.optString("public_did")
        return if (!storedPublicDid.isEmpty()) {
            storedPublicDid == newPublicDid
        } else {
            val newDid = newJson.getJSONArray("recipientKeys").getString(0)
            val storedDid = storedJson.getJSONArray("recipientKeys").optString(0)
            storedDid == newDid
        }
    }

    enum class InvitationType {
        Connection, OutOfBand
    }
}